utils.py

```diff
@@ -935,3 +923,177 @@ def get_latest_folder(folder_path):
     latest = folders[-1]
     logging.info(f"Latest folder: {latest}")
     return latest
+
+
+def is_appimage(file_path, for_argparse=False):
+    # Ref:
+    # - https://cgit.freedesktop.org/xdg/shared-mime-info/commit/?id=c643cab25b8a4ea17e73eae5bc318c840f0e3d4b
+    # - https://github.com/AppImage/AppImageSpec/blob/master/draft.md#image-format
+    # Note:
+    # result is a tuple: (is AppImage: True|False, AppImage type: 1|2|None)
+    result = (False, None)
+    expanded_path = Path(file_path).expanduser().resolve()
+    logging.debug(f"Converting path to expanded_path: {expanded_path}")
+    if file_exists(expanded_path):
+        logging.debug(f"{expanded_path} exists!")
+        with file_path.open('rb') as f:
+            f.seek(1)
+            elf_sig = f.read(3)
+            f.seek(8)
+            ai_sig = f.read(2)
+            f.seek(10)
+            v_sig = f.read(1)
+
+        appimage_check = elf_sig == b'ELF' and ai_sig == b'AI'
+        appimage_type = int.from_bytes(v_sig)
+
+        if appimage_check is False:
+            if for_argparse:
+                raise argparse.ArgumentTypeError(f"Invalid file type: '{file_path}'. Must be of type 'AppImage'.")
+            else:
+                return (appimage_check, appimage_type)
+        return (appimage_check, appimage_type)
+    else:
+        if for_argparse:
+            raise argparse.ArgumentTypeError(f"Invalid file path: '{file_path}'. File does not exist.")
+        else:
+            return (appimage_check, appimage_type)
+
+
+def check_appimage(file_path, for_argparse=False):
+    logging.debug(f"Checking if {file_path} is a usable AppImage.")
+    if file_path is None:
+        logging.error(f"check_appimage: received None for file_path.")
+        return False
+
+    appimage, appimage_type = is_appimage(file_path, for_argparse)
+    if appimage:
+        logging.debug(f"It is an AppImage!")
+        if appimage_type == 1:
+            logging.error(f"{file_path}: Can't handle AppImage version {str(appimage_type)} yet.")
+            return False
+        else:
+            logging.debug(f"It is a usable AppImage!")
+            return True
+    else:
+        logging.debug(f"It is not an AppImage!")
+        return False
+
+
+def find_paths(directory, regex_expression):
+    if directory is None:
+        raise ValueError("Directory path cannot be None.")
+
+    regex_pattern = re.compile(regex_expression, re.IGNORECASE)
+    return [path for path in Path(directory).rglob('*') if regex_pattern.search(str(path))]
+
+
+def find_appimage_files():
+    appimages = []
+    directories = [
+        os.path.expanduser("~") + "/bin",
+        config.APPDIR_BINDIR,
+        get_user_downloads_dir()
+    ]
+    if config.CUSTOMBINPATH is not None:
+        directories.append(config.CUSTOMBINPATH)
+
+    for dir in directories:
+        appimage_paths = find_paths(dir, r'wine.*\.appimage')
```

Comment:
Is this the only place `find_paths` is used? If so, then a simple `Path().rglob('wine*.appimage', case_sensitive=False)` is sufficient, and should be significantly faster than `find_paths`, which checks *every* file and subfolder in `directory` against `regex_expression`.

Reply:
Python kept complaining about this for some reason. I will change it back.